llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -3165,56 +3181,87 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
   return AbbrevEntryReader(AbbrevData);
 }
 
-Error DIEVisitor::visitDIERef(BinaryStreamReader &DataReader,
-                              unsigned AbbrevIdx, StringRef DIEData,
-                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
-  AbbrevEntryReader AbbrevReader =
-      getAbbrevEntryReader(AbbrevEntries, AbbrevIdx);
+static void
+restoreStack(BinaryStreamReader &Reader, StringRef &Data,
+             std::stack<std::pair<StringRef, unsigned>> &StackOfNodes) {
+  auto DataAndOffset = StackOfNodes.top();
+  Reader = BinaryStreamReader(DataAndOffset.first, llvm::endianness::little);
+  Data = DataAndOffset.first;
+  Reader.setOffset(DataAndOffset.second);
+  StackOfNodes.pop();
+}
 
-  if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
-    StartTagCallback(*MaybeTag, AbbrevIdx);
-  else
-    return MaybeTag.takeError();
+Error DIEVisitor::visitDIERef(unsigned AbbrevIdx,
+                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
 
-  Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
-  if (!MaybeHasChildren)
-    return MaybeHasChildren.takeError();
+  std::stack<std::pair<StringRef, unsigned>> StackOfNodes;
+  auto Data = DIEChildrenStack.empty() ? StringRef()
+                                       : DIEChildrenStack.front().getData();
+  BinaryStreamReader Reader(Data, llvm::endianness::little);
+  while (true) {
```

Comment:
these doubly nested `while true` loops are throwing me off, and there is some code duplication between them. Can you provide some intuition why one loop is not enough?

Reply:
You need the inner loop because you could have a `EndOfDIESiblingsMarker` followed by a `DIEInAnotherBlockMarker` followed by the AbbrevIdx, and writing a bunch of nested if's would just be ugly

Reply:
I guess the outer loop could be changed from a `while(true)` to a `while(!DistinctReader.empty())`